Steam - Project 2

GaxysGalactic opened the pull request:
This contains the corrections to the previous project pull request as well as the references added thanks to this project.

GaxysGalactic committed:
Part 1 - Game References

o Changed copies to references in constructor to avoid unwanted copies

GaxysGalactic committed:
Part 2 - Category References

o Changed the GetGame method so it returns a boolean and has an out parameter for the game.

GaxysGalactic committed:
Part 3 - Container & Steam References

o Changed some methods to use references to avoid unnecessary copies
o Changed ValidateInt so that it uses references instead of the awkward assignment and parameter line used earlier.

GaxysGalactic committed:
Final fixes

o Changed CreateGame back, to simply return the game, as it will always be validated.

GaxysGalactic committed:
Copy Constructors and Destructors

+ Added Copy Constructors for each class
+ Added Destructors for each class

GaxysGalactic committed:
Copy Constructor Fix

o Fixed copying arrays for the copy constructors.

GaxysGalactic committed:
Additional References

o Changed methods to include additional references to avoid even more unnecessary copies.

camodesigns reviewed (changes requested):
Keep in mind that not all months have 31 days and that you should not admit dates above the current year since it would not be logical to say that it is valid

JDAlzate commented on Projects/Steam/Steam/Game.h:
```diff
@@ -17,12 +17,16 @@ class FGame
 
 	FGame();
+	
+	FGame(const FGame& oldGame);
```
Careful here with PascalCase.

Also, OldGame is not a very good name here, since you're not actually replacing one game with another, you're just copying one's values into the other. In that case, maybe something like Other or OtherGame may be more appropriate

JDAlzate commented on Projects/Steam/Steam/Steam.cpp:
```diff
@@ -58,7 +73,7 @@ void FSteam::MainMenu()
 	std::cin >> Option;
 
-	Option = ValidateInt(Option, 1, 4);
+	ValidateInt(Option, 1, 4);
```
This method has been working fine, but I have two comments:
1. Why is it necessary to have a validator rather than a getter? Instead of `ValidateInt()` you could have a `GetValidIntInput()` or something like that. That way, the first `cin >> Option` can occur inside the method itself and so you don't have to do it every time you want to get an int input.

2. Ouput parameters are useful in some scenarios, like for example when you want to return a value but also return whether or not the returned value is valid, so you use a bool return type and an out parameter for the other type you want to return.

However, they are not the standard way of returning stuff, that's what the return type is for. With this in mind, although this method works fine, I believe that it would be clearer if you used the return type, the way you were doing before.

Ouput parameters should be special cases, rather than the rule. Especially with built-in types.

JDAlzate commented on Projects/Steam/Steam/Steam.cpp:
```diff
@@ -73,7 +88,7 @@ void FSteam::AddGame()
 {
 	const int CategoryChosen = ChooseCategory();
 
-	const FGame NewGame = CreateGame();
```
Any reason you removed this const here?

JDAlzate reviewed (commented):
Very nice! But please don't forget PascalCase 👀

GaxysGalactic committed:
Pull Request Corrections

+ Added ValidateDayInput method
o Changed some PascalCase issues
o Changed "OldX" to "OtherX" in copy constructors.
